pytorch_lightning/utilities/types.py

```diff
@@ -23,7 +23,9 @@
 from torch.optim import Optimizer
 from torch.utils.data import DataLoader
 from torchmetrics import Metric
-from typing_extensions import Protocol, runtime_checkable, TypedDict
+from typing_extensions import TypedDict
+
+from pytorch_lightning.core.hooks import Stateful
```

Comment:
I strongly suggest that `Stateful` stays in this file. You'll notice there was no `from pytorch_lightning` import here and adding one opens the door for circular dependency hell, as many other source files import `types.py`

Reply:
You can see this already with the red CI

Reply:
@carmocca @jjenniferdai  what if we added this to a new file: core/protocols.py ?

Reply:
I don't think that would work because of the imports in `pytorch_lightning/core/__init__.py`

What's the downside of keeping it here?

What could work too is `pytorch_lightning/protocols.py`. But then we'd have to move this file too for consistency?

Reply:
I see, then moving the code around / mapping these dependencies should be handled in a separate issue/PR. The important thing now is establishing this protocol and updating the various lightning components to adopt it